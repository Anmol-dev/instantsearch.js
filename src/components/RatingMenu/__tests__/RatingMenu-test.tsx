/** @jsx h */

import { h } from 'preact';
import { render, fireEvent } from '@testing-library/preact';
import RatingMenu from '../RatingMenu';

const cssClasses = {
  root: 'root',
  noRefinementRoot: 'noRefinementRoot',
  list: 'list',
  item: 'item',
  selectedItem: 'selectedItem',
  disabledItem: 'disabledItem',
  starIcon: 'starIcon',
  fullStarIcon: 'fullStarIcon',
  emptyStarIcon: 'emptyStarIcon',
  label: 'label',
  count: 'count',
};

const templates = {
  item: `{{#count}}<a class="{{cssClasses.link}}" aria-label="{{value}} & up" href="{{href}}">{{/count}}{{^count}}<div class="{{cssClasses.link}}" aria-label="{{value}} & up" disabled>{{/count}}
  {{#stars}}<svg class="{{cssClasses.starIcon}} {{#.}}{{cssClasses.fullStarIcon}}{{/.}}{{^.}}{{cssClasses.emptyStarIcon}}{{/.}}" aria-hidden="true" width="24" height="24">
    {{#.}}<use xlink:href="#ais-RatingMenu-starSymbol"></use>{{/.}}{{^.}}<use xlink:href="#ais-RatingMenu-starEmptySymbol"></use>{{/.}}
  </svg>{{/stars}}
  <span class="{{cssClasses.label}}">& Up</span>
  {{#count}}<span class="{{cssClasses.count}}">{{#helpers.formatNumber}}{{count}}{{/helpers.formatNumber}}</span>{{/count}}
{{#count}}</a>{{/count}}{{^count}}</div>{{/count}}`,
};

const defaultProps = {
  createURL: value => `#${value}`,
  cssClasses,
  items: [],
  templateProps: {
    templates,
  },
  toggleRefinement: () => {},
  suit: () => '',
};

describe('RatingMenu', () => {
  test('renders correct DOM without items', () => {
    const props = {
      ...defaultProps,
    };

    const { container } = render(<RatingMenu {...props} />);
    const list = container.querySelector('.list');

    expect(list).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('renders correct DOM with items', () => {
    const items = [
      {
        name: '0',
        value: '0',
        count: 3422,
        isRefined: false,
        stars: [false, false, false, false, false],
      },
      {
        name: '1',
        value: '1',
        count: 156,
        isRefined: false,
        stars: [true, false, false, false, false],
      },
      {
        name: '2',
        value: '2',
        count: 194,
        isRefined: false,
        stars: [true, true, false, false, false],
      },
      {
        name: '3',
        value: '3',
        count: 1622,
        isRefined: false,
        stars: [true, true, true, false, false],
      },
      {
        name: '4',
        value: '4',
        count: 13925,
        isRefined: false,
        stars: [true, true, true, true, false],
      },
      {
        name: '5',
        value: '5',
        count: 2150,
        isRefined: true,
        stars: [true, true, true, true, true],
      },
    ];
    const props = {
      ...defaultProps,
      items,
    };

    const { container } = render(<RatingMenu {...props} />);
    const list = container.querySelector('.list');
    const listItems = container.querySelectorAll('.item');

    expect(list).toBeInTheDocument();
    expect(listItems).toHaveLength(items.length);
    expect(container).toMatchSnapshot();
  });

  test('calls templates with correct data', () => {
    const itemTemplate = jest.fn(() => '');
    const props = {
      ...defaultProps,
      items: [
        {
          name: '0',
          value: '0',
          count: 3422,
          isRefined: false,
          stars: [false, false, false, false, false],
        },
      ],
      templateProps: {
        templates: {
          item: itemTemplate,
        },
      },
    };

    render(<RatingMenu {...props} />);

    expect(itemTemplate).toHaveBeenCalledTimes(1);
    expect(itemTemplate).toHaveBeenCalledWith({
      count: 3422,
      cssClasses,
      isRefined: false,
      name: '0',
      stars: [false, false, false, false, false],
      url: '#0',
      value: '0',
    });
  });

  test('triggers refinement on click', () => {
    const toggleRefinement = jest.fn();
    const props = {
      ...defaultProps,
      items: [
        {
          name: '0',
          value: '0',
          count: 3422,
          isRefined: false,
          stars: [false, false, false, false, false],
        },
      ],
      toggleRefinement,
    };

    const { container } = render(<RatingMenu {...props} />);
    const item = container.querySelector('.item')!;

    fireEvent.click(item);

    expect(toggleRefinement).toHaveBeenCalledTimes(1);
    expect(toggleRefinement).toHaveBeenCalledWith('0');
  });
});
