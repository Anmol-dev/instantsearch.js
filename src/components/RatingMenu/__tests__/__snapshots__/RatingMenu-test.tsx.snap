// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingMenu renders correct DOM with items 1`] = `
<div>
  <div
    class="root"
  >
    <svg
      style="display: none;"
      xmlns="http://www.w3.org/2000/svg"
    >
      <symbol
        id=""
        viewBox="0 0 24 24"
      >
        <path
          d="M12 .288l2.833 8.718h9.167l-7.417 5.389 2.833 8.718-7.416-5.388-7.417 5.388 2.833-8.718-7.416-5.389h9.167z"
        />
      </symbol>
      <symbol
        id=""
        viewBox="0 0 24 24"
      >
        <path
          d="M12 6.76l1.379 4.246h4.465l-3.612 2.625 1.379 4.246-3.611-2.625-3.612 2.625 1.379-4.246-3.612-2.625h4.465l1.38-4.246zm0-6.472l-2.833 8.718h-9.167l7.416 5.389-2.833 8.718 7.417-5.388 7.416 5.388-2.833-8.718 7.417-5.389h-9.167l-2.833-8.718z"
        />
      </symbol>
    </svg>
    <ul
      class="list"
    >
      <li
        class="item"
      >
        <div>
          <a
            aria-label="0 & up"
            class=""
            href=""
          >
             
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
             
            <span
              class="label"
            >
              & Up
            </span>
             
            <span
              class="count"
            />
             
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            aria-label="1 & up"
            class=""
            href=""
          >
             
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
             
            <span
              class="label"
            >
              & Up
            </span>
             
            <span
              class="count"
            />
             
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            aria-label="2 & up"
            class=""
            href=""
          >
             
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
             
            <span
              class="label"
            >
              & Up
            </span>
             
            <span
              class="count"
            />
             
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            aria-label="3 & up"
            class=""
            href=""
          >
             
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
             
            <span
              class="label"
            >
              & Up
            </span>
             
            <span
              class="count"
            />
             
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            aria-label="4 & up"
            class=""
            href=""
          >
             
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon emptyStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starEmptySymbol"
              />
               
            </svg>
             
            <span
              class="label"
            >
              & Up
            </span>
             
            <span
              class="count"
            />
             
          </a>
        </div>
      </li>
      <li
        class="item selectedItem"
      >
        <div>
          <a
            aria-label="5 & up"
            class=""
            href=""
          >
             
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
            <svg
              aria-hidden="true"
              class="starIcon fullStarIcon"
              height="24"
              width="24"
            >
               
              <use
                xlink:href="#ais-RatingMenu-starSymbol"
              />
               
            </svg>
             
            <span
              class="label"
            >
              & Up
            </span>
             
            <span
              class="count"
            />
             
          </a>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`RatingMenu renders correct DOM without items 1`] = `
<div>
  <div
    class="root noRefinementRoot"
  >
    <svg
      style="display: none;"
      xmlns="http://www.w3.org/2000/svg"
    >
      <symbol
        id=""
        viewBox="0 0 24 24"
      >
        <path
          d="M12 .288l2.833 8.718h9.167l-7.417 5.389 2.833 8.718-7.416-5.388-7.417 5.388 2.833-8.718-7.416-5.389h9.167z"
        />
      </symbol>
      <symbol
        id=""
        viewBox="0 0 24 24"
      >
        <path
          d="M12 6.76l1.379 4.246h4.465l-3.612 2.625 1.379 4.246-3.611-2.625-3.612 2.625 1.379-4.246-3.612-2.625h4.465l1.38-4.246zm0-6.472l-2.833 8.718h-9.167l7.416 5.389-2.833 8.718 7.417-5.388 7.416 5.388-2.833-8.718 7.417-5.389h-9.167l-2.833-8.718z"
        />
      </symbol>
    </svg>
  </div>
</div>
`;
